File: src/agent/agent-run.ts
import { modelId } from "../config.js";
import { DEFAULT_MAX_TOKENS } from "../constants/llm.js";
import type { CompactionPolicy } from "../context/composer/compaction/policy.js";
import type { ComposedLLMRequest } from "../context/composer/types.js";
import type { SessionStores } from "../session/stores/index.js";
import { extractText } from "../llm/normalize/extract-text.js";
import type { LLMResponse } from "../llm/protocol/types.js";
import { getToolDefinitions } from "../tools/index.js";
import type { Session } from "../session/session.js";
import type { LoopContext } from "./deps.js";
import { composeForSession } from "./compose-for-turn.js";
import { patchLastManifestDeepTask, publishComposeResult } from "./context-status.js";
import {
  getActiveWorkMemId,
  isDeepModeEnabled,
} from "./deep-mode.js";
import {
  ORIENT_PROTOCOL_REMINDER_TEXT,
  SYNTHESIZE_PROTOCOL_REMINDER_TEXT,
  shouldSendOrientProtocolReminder,
} from "./deep-task-protocol.js";
import { getWorkMemAgentPorts } from "./ports/work-mem.js";
import { runLLM } from "./pipeline/runLLM.js";
import { runToolUses } from "./pipeline/runTool.js";

export interface AgentRunComposeOptions {
  resumeFromCheckpointId?: string;
  activeCompactionSaveId?: string;
  getCompactionPolicy: () => CompactionPolicy;
  onAfterCompose?: () => void;
}

export class AgentRun {
  private readonly session: Session;
  private readonly stores: SessionStores;
  private readonly loopCtx: LoopContext;
  private readonly composeOptions: AgentRunComposeOptions;
  private orientProtocolReminderSent = false;
  private synthesizeProtocolReminderSent = false;

  constructor(
    session: Session,
    stores: SessionStores,
    loopCtx: LoopContext,
    composeOptions: AgentRunComposeOptions,
  ) {
    this.session = session;
    this.stores = stores;
    this.loopCtx = loopCtx;
    this.composeOptions = composeOptions;
  }

  async execute(userPrompt: string): Promise<{ reply: string; turn: number }> {
    const { runtime } = this.loopCtx;
    await runtime.hooks.dispatch("runStart", { userPrompt });
    try {
      await this.recordUser(1, userPrompt);
      let runTurn = 0;
      while (true) {
        runTurn += 1;
        const input = await this.buildInput(runTurn);
        const response = await this.callModel(runTurn, input);
        const done = await this.recordOutcome(runTurn, response);
        if (done) {
          const result = { reply: done.reply, turn: runTurn };
          await runtime.hooks.dispatch("runEnd", result);
          return result;
        }
      }
    } catch (error) {
      await runtime.hooks.dispatch("runError", { error });
      throw error;
    } finally {
      await runtime.hooks.dispatch("runFinalize", {});
    }
  }

  private async recordUser(turn: number, prompt: string): Promise<void> {
    await this.session.appendUser(turn, prompt);
  }

  private async buildInput(runTurn: number): Promise<ComposedLLMRequest> {
    const composed = await composeForSession({
      session: this.session,
      stores: this.stores,
      turn: runTurn,
      toolDefinitions: getToolDefinitions(this.loopCtx.runtime.tools),
      compactionPolicy: this.composeOptions.getCompactionPolicy(),
      resumeFromCheckpointId: this.composeOptions.resumeFromCheckpointId,
      activeCompactionSaveId: this.composeOptions.activeCompactionSaveId,
    });
    publishComposeResult(composed);
    await this.loopCtx.runtime.hooks.dispatch("composeComplete", { composed });
    this.composeOptions.onAfterCompose?.();
    return composed.request as ComposedLLMRequest;
  }

  private async callModel(runTurn: number, input: ComposedLLMRequest) {
    return runLLM({
      turn: runTurn,
      runtime: this.loopCtx.runtime,
      model: modelId(),
      maxTokens: DEFAULT_MAX_TOKENS,
      messages: input.messages,
      system: input.system,
      tools: input.tools,
    });
  }

  private async recordOutcome(
    runTurn: number,
    response: LLMResponse,
  ): Promise<{ reply: string } | undefined> {
    await this.session.appendAssistant(runTurn, response.content);

    if (response.stopReason === "tool_use") {
      if (
        isDeepModeEnabled()
        && runTurn === 1
        && !this.orientProtocolReminderSent
        && shouldSendOrientProtocolReminder(response.content)
      ) {
        await runToolUses(runTurn, response.content, this.loopCtx);
        await this.session.appendProtocolReminder(runTurn, "orient", ORIENT_PROTOCOL_REMINDER_TEXT);
        this.orientProtocolReminderSent = true;
        return undefined;
      }

      await runToolUses(runTurn, response.content, this.loopCtx);
      return undefined;
    }

    if (isDeepModeEnabled() && !this.activeWorkMemHasDecision()) {
      if (!this.synthesizeProtocolReminderSent) {
        await this.session.appendProtocolReminder(
          runTurn,
          "synthesize",
          SYNTHESIZE_PROTOCOL_REMINDER_TEXT,
        );
        this.synthesizeProtocolReminderSent = true;
        return undefined;
      }
      patchLastManifestDeepTask({ synthesizeSkipped: true });
    }

    return { reply: extractText(response.content) };
  }

  private activeWorkMemHasDecision(): boolean {
    const workMemId = getActiveWorkMemId(this.session.sessionId);
    if (!workMemId) {
      return false;
    }
    return getWorkMemAgentPorts().hasDecisionDraft({
      sessionId: this.session.sessionId,
      workMemId,
    });
  }
}

File: crates/ocula-agent/src/agent_run.rs
use std::path::{Path, PathBuf};
use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;
use ocula_composer::compose_context_v1;
use ocula_llm::{extract_text, LlmClient};
use ocula_protocol::ContentBlock;
use ocula_session::Session;
use ocula_tools::UserInteraction;

use crate::loop_config::LoopConfig;
use crate::prompt::build_system_prompt;
use crate::tool_pipeline::run_tool_uses;

pub struct LoopContext {
    pub workdir: PathBuf,
    pub session: Session,
    pub interaction: Arc<dyn UserInteraction>,
    pub llm: Arc<dyn LlmClient>,
    pub loop_config: Arc<dyn LoopConfig>,
}

pub struct AgentRun {
    ctx: LoopContext,
}

impl AgentRun {
    pub fn new(ctx: LoopContext) -> Self {
        Self { ctx }
    }

    pub async fn execute(&self, user_prompt: &str) -> Result<RunResult> {
        self.ctx.session.append_user(1, user_prompt).await?;

        let mut run_turn = 0u32;
        loop {
            run_turn += 1;
            let system = build_system_prompt(&self.ctx.workdir);
            let slice = self.ctx.session.log_slice().await?;
            let mut messages = slice.to_messages(None);
            messages = self.ctx.loop_config.transform_context(messages);

            let composed = compose_context_v1(system, messages);
            let response = self
                .ctx
                .llm
                .chat(&composed.messages, &composed.tools, &composed.system)
                .await?;

            self.ctx
                .session
                .append_assistant(run_turn, response.content.clone())
                .await?;

            if response.stop_reason != "tool_use" {
                return Ok(RunResult {
                    reply: extract_text(&response.content),
                    turn: run_turn,
                });
            }

            run_tool_uses(
                &response.content,
                run_turn,
                &self.ctx.session,
                &self.ctx.workdir,
                self.ctx.interaction.clone(),
            )
            .await?;
        }
    }
}

pub struct RunResult {
    pub reply: String,
    pub turn: u32,
}
